step-plans/step-plans-core/src/main/java/step/core/execution/model/ExecutionAgentUrls.java

```diff
@@ -0,0 +1,39 @@
+package step.core.execution.model;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * Simple data class to keep track of which agents (identified by their URLs)
+ * participated in an execution.
+ */
+public class ExecutionAgentUrls {
+    private final Set<String> backingSet = ConcurrentHashMap.newKeySet();
+
+    public void add(String agentUrl) {
+        backingSet.add(agentUrl);
+    }
+
+    @Override
+    // this directly returns the data in the output format the frontend expects:
+    // space-separated, and for convenience, sorted alphabetically (case-insensitive)
+    public String toString() {
+        return backingSet.stream()
+                .sorted(String.CASE_INSENSITIVE_ORDER)
+                .collect(Collectors.joining(" "));
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == null || getClass() != o.getClass()) return false;
```

Comment:
Where do we need this method?

Reply:
That was just for good practice, using generate hashCode and equals to follow the usual java contract. Will probably be gone in next iteration.